### ISSUE TITLE
narrowing regression for dataclass instances (1.1.380)

### ISSUE THREAD

**Issue by catwell** [microsoft/pyright#8963]
**Describe the bug**

Pyright 1.1.380 incorrectly narrows the type of dataclass instances when using `isinstance(..., type)`, which is [how it should be done according to the Python documentation](https://docs.python.org/3/library/dataclasses.html#dataclasses.is_dataclass).

I suspect this is due to this:

> Fixed a bug that leads to incorrect type narrowing with `isinstance` or `TypeIs` when the filter type (the second argument to `isinstance`) is a subclass of `type`.

**Code or Screenshots**

```python
import dataclasses as dc
from typing import Any

def p(o: Any) -> None:
  if dc.is_dataclass(o) and not isinstance(o, type):
    print(dc.asdict(o))
```

With 1.1.379 this (correctly) type checks. With 1.1.380 this returns:

```
foo.py:6:21 - error: Argument of type "DataclassInstance | type[DataclassInstance]" cannot be assigned to parameter "obj" of type "DataclassInstance" in function "asdict"
    Type "DataclassInstance | type[DataclassInstance]" is not assignable to type "DataclassInstance"
      "__dataclass_fields__" is defined as a ClassVar in protocol (reportArgumentType)
1 error, 0 warnings, 0 informations
```

**Comment by erictraut** [comment]
This is addressed in pyright 1.1.381.
